hcl.html.markdown

```diff
@@ -0,0 +1,255 @@
+---
+category: tool
+tool: hcl
+contributors:
+    - ["Romans Malinovskis" , "http://github.com/romaninsh"]
+filename: terraform.txt
+---
+
+## Introduction
+
+HCL (Hashicorp Configuration Language) is a high-level configuration language used in tools from 
+Hashicorp (such as Terraform). This document focuses on a most recent HCL syntax (0.12).
+
+HCL is a declarative language and will consume all `*.tf` in the current folder. Sub-folders
+can be consumed through modules.
+
+This guide is focused on HCL specifics, you should be already familiar with what is Terraform.
+
+```hcl-terraform
+// Top-level HCL file will interactively ask user values for the variables
+// which do not have default value
+variable "ready" {
+  description = "Ready to learn?"
+  type = bool
+  // default = true
+}
+
+// Module block consults a specified folder for *.tf files, would
+// effectively prefix all resources id's with "module.learn-basics."
+// 
+module "learn-basics" {
+  source = "./learn-basics"
+  ready_to_learn = var.ready
+}
+
+output "knowledge" {
+  value = module.learn-basics.knowledge
+}
+```
+
+## learn-basics
+
+```hcl-terraform
+
+// Variables are not automatically passed into modules
+// and can be typeless.
+variable "ready" {
+}
+
+// It is good practice to define type though. There are 3 primitive types
+// 3 collection types and 2 structural types. Structural types define
+// types recursively
+variable "structural-types" {
+  type = object({
+    object: object({
+      can-be-nested: bool
+    }),
+    tuple: tuple([int, string])
+  })
+  
+  default = {
+    object = { can-be-nested: true }
+    tuple = [3, "cm"]
+  }
+}
+
+// Collection types may specify type, but can also be "any".
+variable "list" {
+  type: list(string)
+  default = ["red", "green", "blue"]
+}
+
+variable "map" {
+  type: map(any)
+  default = {
+    red = "#FF0000"
+    "green" = "#00FF00"
+  }
+}
+
+variable "favourites" {
+  type: set
+  default = ["red", "blue"]
+}
+
+// When type is not specified or is mixed of scalars they will be
+// converted to strings.
+
+
+// IDEs (such as IntelliJ) syntax highlighting
```

Comment:
while I like and use IDEs from Jetbrains, I don't think it makes sense to include this reference in the LXIYM guide